### PR TITLE
Particle spawner: Do not spawn particles distant from player

### PR DESCRIPTION
**Pull request by paramat**
As #5760 but for particle spawners. Second part of fix for #5667 
Tested.

> Not sending particle spawners is a problem since they tend to be longer lived than particles. If I walk into the range of a particle spawner I should be able to see it.
Send the particle spawner always, but just don't generate particles if it's too far away. This will still solve the performance problems because they are caused by particle rendering, not particle spawners themselves.

Acts as requested, a particle spawner always is sent to the client but each individual particle is checked clientside for distance from the player to decide whether it is actually spawned @raymoo 

There was already a conditional to deal with unloaded parents for attached spawners so i added the distance check to that conditional to save another indentation, the distance check comes first as a player being distant is more likely than an unloaded parent.

@Fixer-007 please could you test? However this is probably not for the release so not a priority.

### CODE REVIEW

**Comment by Fixer-007** [comment]
I've tried some TNT tests on localhost, looks like this does not help, I have noticeable fps drop on distant client during TNT explosion. Will check on hometown too.

**Comment by nerzhul** [comment]
it's not the good solution, the good solution is to add/remove particlespawner when a client is near from it from server side, but it require more work and should be added after release.

**Comment by raymoo** [comment]
@nerzhul If particles aren't generated then there shouldn't be draw lag.
@Fixer-007 Can you tell if particles are still being generated?

Having the server send/remove particlespawners is good for other reasons though, like if a client joins after a particlespawner is added.

**Comment by Fixer-007** [comment]
>Can you tell if particles are still being generated?
Don't understand you. Explosion was very far away, my particle meter detected them, but I'm not sure if they were rendered itself.

**Comment by paramat** [comment]
> This patch is client-side, right?

Yes.

>  Does hometown need "Particles: Do not send single particles to distant clients (#5760)" applied?

No, if you test with a mod that only uses particle spawners then that PR is not needed.

Thanks, weird that it does not work, something for after release.

**Comment by paramat** [comment]
Oops:
`if (pos.getDistanceFrom(ppos) <= radius || !unloaded) {`
Should be '&&'.

**Comment by paramat** [comment]
Updated. Will now test.

> it's not the good solution, the good solution is to add/remove particlespawner when a client is near from it from server side, but it require more work and should be added after release.

I disagree, that will be complex to do and most spawners are short-lived anyway so dynamically adding and removing them with player distance is not much help because the timespan of a player moving in and out of range is usually larger than the lifespan of the spawner.
It also requires storing the particle spawner when disabled but still dummy-running the spawning of particles because a spawner has a fixed duration and a fixed particle rate. This would be little different to how this PR dummy-runs the spawning of distant particles.

So this PR is simple and assuming it works in testing i feel should be in release to fix this very severe reduction in performance, we can always use a different implementation later if we decide to, but this is more than good enough for a release.

**Comment by paramat** [comment]
Tested and fine. Additional testing by @Fixer-007 would be good if possible.

**Comment by Fixer-007** [comment]
Will retest, including on updated hometown.

**Comment by Fixer-007** [comment]
This patch is client side only? Or needs apply on server too?

**Comment by nerzhul** [comment]
client side only, server side require more engineering

**Comment by Fixer-007** [comment]
Looks fine under localhost. I wonder what are those particles in my base on hometown server, is it possible to get their coordinates?

File: src/particles.cpp [hunk 1]
```diff
@@ -316,11 +320,12 @@ void ParticleSpawner::step(float dtime, ClientEnvironment* env)
 			{
 				m_amount--;
 
-				// Pretend to, but don't actually spawn a
-				// particle if it is attached to an unloaded
-				// object.
-				if (!unloaded) {
```
Comment (nerzhul) [inline src/particles.cpp]: as there are more useless calculs when !unloaded, please preserve !unloaded as a parent condition and instance your variables inside it

**Comment by numberZero** [comment]
@Fixer-007 surely that’s possible. I suggest adding printing (to `std::cout`, `verbosestream` or whatever) to `ParticleManager::addParticle` (`src/particles.cpp`)

**Comment by Fixer-007** [comment]
Any help on this? :D

File: src/particles.cpp (hunk 1 above)
Comment (paramat) [inline src/particles.cpp]: Ok.

Comment (paramat) [inline src/particles.cpp]: Done.

**Comment by paramat** [comment]
Updated as requested  and retested.

Fixer thanks for testing.

**Comment by paramat** [comment]
Ugh just realised i have not distance-limited infinite particle spawners, will fix.

**Comment by paramat** [comment]
Updated but not yet tested.

**Comment by VanessaE** [comment]
@paramat if particles cause FPS lag (which they do), fix the particles, not the mod that uses them... and they DO default to "off".   You have to right-click on a faucet to make it produce particles, and I made them as light as I could while still looking like water flow.

**Comment by VanessaE** [comment]
Also remember, jin_xi had some well-working code that made the engine use proper irrlicht particles.  10'000 of them on the screen had only a small effect on my FPS, and I was using a mid-range GPU at the time.

**Comment by nerzhul** [comment]
yeah @VanessaE it should be ported and tested but too late for this release cycle

**Comment by paramat** [comment]
Since particles are intensive, a mod has to take this into account and not overuse them. By default off i mean disable having water when you turn them on, and provide a warning in the mod about the load caused by enabling.
Of course we will need to work on particles, seems top priority at the moment. The crazy thing i am fixing here is an example of how bad they are.

**Comment by VanessaE** [comment]
For sure, it needs ported over, but until then, steps need to be taken in-engine to limit the FPS lag - maybe just have the engine automatically reduce the number of displayed particles if the FPS drops by more than, oh, 5 percent or so.

**Comment by paramat** [comment]
No that's an over-complex suggestion, if particles become too heavy the mos can be removed or edited.

**Comment by paramat** [comment]
Retested on a local server and fine.

**Comment by VanessaE** [comment]
Honestly, removing a mod because particles are slow is a terrible solution.  Make the particles fast, whatever it takes, and make the engine limit how many it displays.  Maybe just allow for a static limit in the user's config.

**Comment by paramat** [comment]
No, limiting particles on player FPS is an overkill complexity suggestion, you do not have to write and maintain it, it's different when you are a dev and realise how little time we have. If particles are a problem the server should edit that mod or not use it.

**Comment by ExeterDad** [comment]
So with that logic we shouldn't have TNT. That's going to be a popular proposal.

**Comment by paramat** [comment]
I do not imply that at all. if TNT particles are a problem then either contact us to suggest improvements, or ask for a parameter to alter number of particles, or edit the mod yourself to have a special version for your server. All these are simple approaches.

**Comment by VanessaE** [comment]
@paramat  "or ask for a parameter to alter number of particles"  um, I *just* said that.

**Comment by paramat** [comment]
> contact us to suggest improvements, or ask for a parameter to alter number of particles,

Sorry, here i mean improvements for the mod, or a parameter in the mod.

But yes, a clientside solution would be needed due to clients varying machine power, so i would suggest making the 'particles' setting in the basic menu affect all particles (as i thought it did) not just digging particles.
Anyway best discussed in a separate issue.

**Comment by Fixer-007** [comment]
Tested it on localhost server with edited rain mod, tnt, and time=0 particles - seems fine, hometown is fine too (had no increases to 300+ particles anymore, so limiter works).

**Pull request merged**